airbyte-cdk/python/airbyte_cdk/sources/declarative/requesters/paginators/strategies/stop_condition.py

```diff
@@ -0,0 +1,42 @@
+#
+# Copyright (c) 2023 Airbyte, Inc., all rights reserved.
+#
+
+from abc import ABC, abstractmethod
+from typing import Any, List, Optional
+
+import requests
+from airbyte_cdk.sources.declarative.incremental import Cursor
+from airbyte_cdk.sources.declarative.requesters.paginators.strategies.pagination_strategy import PaginationStrategy
+from airbyte_cdk.sources.declarative.types import Record
+
+
+class PaginationStopCondition(ABC):
+    @abstractmethod
+    def is_met(self, record: Record) -> bool:
+        raise NotImplementedError()
+
+
+class CursorStopCondition(PaginationStopCondition):
+    def __init__(self, cursor: Cursor):
+        self._cursor = cursor
+
+    def is_met(self, record: Record) -> bool:
+        return not self._cursor.should_be_synced(record)
+
+
+class StopConditionPaginationStrategyDecorator(PaginationStrategy):
```

Comment:
Just for clarification - this means we will still send the full page even if parts of the page are already meeting the condition, right? Can you explain a bit the reasoning why we aren't evaluating this for every record to stop syncing at exactly the right place? As I imagine we want to add semi-incremental sync the same way which is only there for reducing duplicate records, but this strategy can't completely prevent them.

Not a strong blocker, just interested in what was discussed around this topic.

Reply:
> this means we will still send the full page even if parts of the page are already meeting the condition, right?

Yes

> Can you explain a bit the reasoning why we aren't evaluating this for every record to stop syncing at exactly the right place?

Because we agreed to decouple the semi-incremental and the data feed to deliver value early and what you are suggesting is part of the semi-increment (i.e. "the API can't filter based on the cursor so I want to filter out some records"). If this point was not well understood, I don't mind bumping the priority of semi-incremental

Reply:
Ah right, no I think it fine, it was more about clarification of the plans beyond this pr.